### PR TITLE
fix(nlpgo): stop code blocks aborting at 120s

### LINKED ISSUE #115: Agent `timeoutMs` is accepted by the API, hidden in Studio, and silently clamped away
## The defect

A Code agent's `timeoutMs` can be set, is persisted, is sent all the way to the Go engine — and is then silently discarded if it asks for more time than the deployment ceiling. The user gets no feedback that their value was ignored.

### 1. Studio cannot set it

`CodePropertiesPanel` renders with `hideParameters` (`platform/app/src/optimization_studio/components/properties/CodePropertiesPanel.tsx:150`), which gates the generic FieldsForm off at `BasePropertiesPanel.tsx:696`. `buildAgentParameters` "code" case (`platform/app/src/optimization_studio/utils/agentNodeData.ts:141-153`) only emits the `code` parameter. `codeComponentSchema.timeoutMs` exists at `platform/app/src/optimization_studio/types/dsl.ts:560` with zero UI writers.

### 2. The REST API accepts and persists it

`POST /api/agents` (`platform/app/src/app/api/agents/[[...route]]/app.ts:31,106-125`) takes `config: z.record(z.unknown())`; `agent.repository.ts:59-61,127,140` validates with `codeComponentSchema.parse()`, so `timeoutMs` survives into DB `agent.config`.

### 3. It is wired end-to-end, then clamped down only

`data-prefetcher.ts:945-947,983` reads `config.timeoutMs` -> `code-agent.adapter.ts:308-338` emits node param `timeout_ms` -> Go `engine.go:600,617-618,1546-1560` -> clamped at `services/nlpgo/app/engine/blocks/codeblock/codeblock.go:268-271`:

```go
timeout := e.opts.DefaultTimeout
if req.Timeout > 0 && req.Timeout < timeout {
    timeout = req.Timeout
}
```

`<` only — it can shorten the budget, never lengthen it. A second independent clamp with the same direction lives in `services/nlpgo/app/engine/blocks/blocktimeout/blocktimeout.go`, whose package doc states a node's `timeout_ms` "is a request for a SHORTER budget: it can never buy more than the deployment allows."

(All line refs verified against `langwatch/langwatch` at time of filing; re-confirm before acting, they may have moved.)

## Why it matters

A user hitting "the code block ran past its time limit and was stopped" will reasonably set `timeoutMs` via the API to fix it. The call returns 200, the value is stored, and nothing changes. There is no warning, no log, no surfaced value. It looks like a broken feature rather than a deliberate ceiling.

## Suggested directions (options, not prescriptions)

- Reject or warn at the API when `timeoutMs` exceeds the deployment ceiling.
- Surface the effective clamped value back in the run result.
- Expose it read-only in Studio so the discrepancy is visible.

Note: the clamp direction itself is deliberate and correct — the defect is the silence, not the clamping.

## Related

- https://github.com/langwatch/langwatch/pull/7614 (added the operator-configurable ceilings)
- https://github.com/langwatch/langwatch-saas/pull/1183 (sets the ceiling in prod)

### PR DESCRIPTION
**Pull request by drewdrewthis**
Combines #7643, #7644, #7645 and #7646. One fix, one review surface.

> **Scariest thing** — this raises a client-side network deadline from **120s to 630s**. A genuinely hung code block now occupies a scenario worker for ~10 minutes instead of ~2. It is bounded by the engine's own 600s ceiling and Lambda's 900s `Timeout`, but that is the failure mode to push on.
> **Start here** — `platform/app/src/server/nlpgo/timeouts.ts` (new, 233 lines). Every number in this PR is derived there. If that file is right, the rest follows.

## Why

Closes https://github.com/langwatch/tasks/issues/115

Production scenario runs died at ~121s:

```
SerializedCodeAgentAdapterError: ... did not respond within 120000ms (request aborted)
```

The engine was willing to run the code block for 600s. The client hung up at 120s. Nobody was wrong individually — the two numbers were just written in different files by different people and never introduced.

Follow-ups tracked in https://github.com/langwatch/tasks/issues/115 and https://github.com/langwatch/tasks/issues/116. There is no in-repo issue for the original report, so this PR closes nothing automatically.

## The bug in one table

Six serialized adapters exist. Only two of them call nlpgo, and on `main` **both** hang up at 120s:

| Adapter | Calls nlpgo? | Deadline on `main` | After this PR |
|---|---|---|---|
| `code-agent` (no explicit `timeoutMs`) | yes | **120s** | **630s** |
| `code-agent` (explicit `timeoutMs`) | yes | grows from 120s | 630s |
| `workflow-agent` | yes | **120s** flat | **630s** |
| `http-agent`, `prompt-config`, `connected-agent` | no | n/a | n/a |

The default path — no explicit `timeoutMs` — was capped at 120s on **both** adapters. Raising the engine's ceiling alone (which is what the already-merged SaaS change https://github.com/langwatch/langwatch-saas/pull/1185 does) does **not** fix either default path, because the client still gives up first. That is why this PR is needed on top of 1185, not instead of it.

## What changed

Four decisions, each with the road not taken:

- **Derive the client deadline from the engine's own ceiling** (`600 + 30s headroom = 630s`) → *instead of* adding an `NLP_FETCH_TIMEOUT_MS` env var → a second knob can silently disagree with the engine's ceiling and reintroduce exactly this bug. Deriving keeps one number authoritative and adds **zero net-new env vars**.

- **Clamp operator overrides at 710s** (`min(Lambda 900, nlpgo stream-idle 720) − 10s margin`) → *instead of* honouring anything up to Lambda's 900s → nlpgo tears the stream down at 720s, so a value of 711–890 would fail with an opaque transport error instead of the engine's own timeout message. **Cost:** anyone who deliberately set 711–890 now gets 710.

- **Fold the timeout into `buildDesiredLambdaEnvironmentVariables()` and reconcile on every ARN cache miss, then key the ARN cache on that same desired config** → *instead of* setting it only at Lambda create time, or keying the cache on `image_uri` alone → already-created per-project Lambdas would keep the 120s behaviour forever, and a config-only rollout (no new image) would serve a warm cached ARN and skip the reconcile for up to the 10-minute TTL. **Cost:** one config comparison per ARN resolution, and a one-time cache invalidation on rollout.

- **Fail the Helm render above 710 rather than clamping silently** → *instead of* accepting and quietly lowering → the chart would accept `715` while the client actually ran `710`, so deployed behaviour would not match declared config. **Cost:** an install pinning 711–719 now fails to render until the value changes.

## How it works

```
platform/app/src/server/nlpgo/timeouts.ts        NEW — the only place a timeout number is decided
├── resolveFloorFetchTimeoutMs()   engine ceiling + 30s headroom ....... 630_000 ms
├── resolveMaxFetchTimeoutMs()     hard upper bound .................... 900_000 ms
└── clamp constants                LAMBDA 900 / IDLE 720 / MARGIN 10 ... ceiling 710 s

  consumed by ──┬── serialized-adapters/code-agent.adapter.ts      (deleted its own 120_000)
                ├── serialized-adapters/workflow-agent.adapter.ts  (deleted its own 120_000)
                └── optimization_studio/server/lambda/index.ts     (clampCodeBlockTimeoutSeconds)

  enforced at ──┬── charts/langwatch/templates/_helpers.tpl        render-time guard, refuses > 710
                ├── charts/langwatch/templates/{app,workers,langwatch_nlp}/deployment.yaml
                └── langwatch-saas#1185 (MERGED)                   same var, scenario workers
```

Both adapters previously declared their own `const NLP_FETCH_TIMEOUT_MS = 120_000`. Both declarations are deleted; both now import from `timeouts.ts`.

**Timeout chain, outermost to innermost:**

| Layer | Value | Why |
|---|---|---|
| Lambda `Timeout` | 900s | outermost bound |
| client fetch max | 900s | matches Lambda |
| nlpgo `StreamIdleTimeoutSeconds` | 720s | deliberately above the ceiling — a *silent* code block must not race stream-idle |
| chart / clamp ceiling | 710s | `min(900, 720) − 10` |
| client fetch floor | 630s | `600 + 30s` headroom |
| nlpgo engine ceiling | 600s | code-block hard stop |

`600 < 720` is deliberate, not an inversion.

## Deployment Impact

**Requires deploy:** yes — platform app, plus Helm consumers.
**Net-new env vars:** zero. Everything derives from `NLPGO_ENGINE_CODE_BLOCK_TIMEOUT_SECONDS`, already plumbed to nlpgo.

| Surface | What changes | Action needed |
|---|---|---|
| `platform/app` | new `nlpgo/timeouts.ts`; adapters and lambda read from it | normal app deploy |
| per-project `langwatch_nlp` Lambdas | timeout + `MemorySize` 2048 reconciled on the next ARN cache miss | none — the ARN cache is keyed on desired config, so a config-only rollout forces a miss and reconciles on the next studio request |
| `charts/langwatch` | new `langwatch_nlp.codeBlockTimeoutSeconds`, default 600 | none — the default preserves today's behaviour |
| SaaS terraform | already merged as langwatch-saas#1185 | done |

**Breaking for three narrow groups:**

1. Chart values in **711–719** — previously rendered fine, then ran at 710 anyway. Now the render fails with the derivation in the message. Re-pick ≤ 710.
2. `NLPGO_ENGINE_CODE_BLOCK_TIMEOUT_SECONDS` set in **711–890** — now clamped down to 710.
3. **Non-numeric or negative** chart values (`abc`, `-5`) — Helm's `int` read these as 0 and rendered successfully. Now refused. An explicit `0` still yields 600, because `default 600` cannot distinguish it from unset.

**Rollback:** revert the app deploy. The chart value defaults to the pre-existing 600, so unsetting it is a no-op. Reconciled Lambdas keep the env var until the next reconcile with reverted code removes it.

**Risk:** the client deadline goes 120s → 630s, so a hung block holds a worker longer before failing. Bounded by the 600s engine ceiling and the 900s Lambda `Timeout`.

**Second-order effect of the cache fingerprint:** a config-only rollout now invalidates every project's cached ARN simultaneously, so the first Studio request per project after rollout pays one `GetFunction` + reconcile. This is the same burst an image deploy has always caused — the fingerprint adds a new trigger, not a new mechanism. Single-flight collapses concurrent misses per pod, not cluster-wide. Not sized against current project counts.

## Human verification

 Backend only; no UI surface changes.

Three checks a skeptic can run independently:

1. **The fix works.** Run a scenario whose code block takes >120s and <600s. Before: aborts at ~120s. After: completes.
2. **The deadline is derived, not hardcoded.** Set `NLPGO_ENGINE_CODE_BLOCK_TIMEOUT_SECONDS=30` and re-run. It should abort at ~60s — proving the client floor tracks the engine ceiling per call.
3. **Existing Lambdas are healed, not just new ones.** On an **already-created** per-project function, issue one studio request, then `aws lambda get-function-configuration --function-name <fn>`: `MemorySize` is 2048, the managed env vars match `buildDesiredLambdaEnvironmentVariables`, and unmanaged vars are untouched.

## How I can prove I was successful

Each claim is tagged **proven** (exercised this session, evidence attached), **claimed** (asserted, not independently exercised), or **missing** (no evidence yet).

| Claim | Status | Evidence |
|---|---|---|
| CI is green at `be0c5c3a8e` | **proven** | `verify-ci-shard-tally.sh langwatch/langwatch 7647` → exit 0; all 8 shards printed real tallies (`test-unit=4`, `test-integration=4`), attempt 1. Not a check-name count — a per-shard floor. |
| Timeout unit tests pass | **proven** | `npx vitest run timeouts.unit.test.ts clampCodeBlockTimeoutSeconds.unit.test.ts` → exit 0, **31/31** |
| Helm guard refuses out-of-range values | **proven** | `charts/langwatch/tests/nlpgo-timeout-guard.sh` → exit 0, **14/14** checks, including a boundary case asserting 710 accepted and 711 refused |
| Chart ceiling equals the Lambda clamp | **proven** | Both derive `min(900, 720) − 10 = 710`; constants read at this SHA, not from memory |
| Live AWS reconcile on an existing Lambda | **claimed** | Source-level plus unit tests only — no run against real AWS. The cache-invalidation half is **proven**: reverting the fingerprint check fails `invalidates the cache when the desired timeout changes` with `expected 3 to be greater than 3`. Verification step 3 above is how to confirm the AWS half. |
| End-to-end production behaviour | **claimed** | Not exercised against a running deployment. |
| Helm guard suite runs in CI | **missing** | The suite has **no CI runner** — the workflow hunk could not be pushed (token lacks `workflow` scope). Tracked at https://github.com/langwatch/tasks/issues/115. It passes locally and nothing re-runs it. |

Visual proof is exempt here: backend-only, no UI surface (declared above).

## Follow-ups

- The literal `600` is duplicated across 5+ sites with no shared source — https://github.com/langwatch/tasks/issues/116
- The chart guard suite needs a CI runner — https://github.com/langwatch/tasks/issues/115
- Companion infra change, already merged — https://github.com/langwatch/langwatch-saas/pull/1185

## Summary by CodeRabbit

* **New Features**
  * Added configurable NLP code-block execution timeouts, defaulting to 600 seconds.
  * Added configurable maximum NLP response times with safer handling of invalid settings.
  * Improved automatic synchronization of project execution environments, including memory and environment configuration.

* **Bug Fixes**
  * Prevented timeout settings from exceeding the engine’s idle limit.
  * Improved handling of concurrent environment updates.
  * Aligned workflow and code-agent timeouts with configured NLP execution limits.

### CODE REVIEW

File: charts/langwatch/templates/langwatch_nlp/deployment.yaml [hunk 1]
```diff
@@ -119,6 +119,15 @@ spec:
...
             {{- end }}
 
+            # Ceiling on how long a workflow code block may run before nlpgo
+            # kills it. Kept below the engine's stream idle timeout by default
+            # so the code-block ceiling can't race the stream shutting down.
+            {{- if ge (int (.Values.langwatch_nlp.codeBlockTimeoutSeconds | default 600)) 720 }}
+            {{- fail "langwatch_nlp.codeBlockTimeoutSeconds must stay below the engine's stream idle timeout (720s) — a code-block ceiling at or above it races the stream shutting down" }}
+            {{- end }}
+            - name: NLPGO_ENGINE_CODE_BLOCK_TIMEOUT_SECONDS
+              value: {{ .Values.langwatch_nlp.codeBlockTimeoutSeconds | default 600 | quote }}
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply charts/langwatch/templates/langwatch_nlp/deployment.yaml]: Fixed — but not at the line this thread is anchored to, which is why it did not go outdated.

The propagation landed in `charts/langwatch/templates/_helpers.tpl` inside the `langwatch.sharedEnv` define (line 707), which `app/deployment.yaml:140` and `workers/deployment.yaml:137` both include. Verified by render rather than by inspection:

```
$ helm template verify charts/langwatch \
    --set langwatch_nlp.codeBlockTimeoutSeconds=700 \
    --set autogen.enabled=true --set postgresql.chartManaged=true \
    --set postgresql.auth.database=langwatch --set redis.chartManaged=true \
    --set clickhouse.chartManaged=true

SRC=langwatch/templates/app/deployment.yaml          value: "700"
SRC=langwatch/templates/langwatch_nlp/deployment.yaml value: "700"
SRC=langwatch/templates/workers/deployment.yaml      value: "700"
```

Default render (no `--set`) emits `"600"` on all three, so the `| default 600 | quote` behavior is preserved. One operator-set number is now read on both sides of the socket.

File: specs/nlp-go/studio-lambda-cache.feature [hunk 2]
```diff
@@ -83,3 +83,46 @@ Feature: getProjectLambdaArn — per-project ARN cache + single-flight
...
+    When getProjectLambdaArn resolves that project
+    Then UpdateFunctionCode is called and polled to completion
+    And only then is UpdateFunctionConfiguration called
+
+  @integration @unit
+  Scenario: A concurrent update makes AWS reject the reconcile but resolution still succeeds
+    Given a Lambda function exists with a drifted MemorySize
+    And AWS rejects the UpdateFunctionConfiguration call with "An update is in progress"
+    When getProjectLambdaArn resolves that project
+    Then the resolution still succeeds and returns a valid ARN
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply specs/nlp-go/studio-lambda-cache.feature:142]: Acknowledged, and deliberately not changing this one.

The scenario prose names `ResourceConflictException` and the Lambda config fields on purpose: this feature file is the executable spec for the reconcile path, and the AWS identifiers are the contract under test, not incidental implementation detail. Genericising the prose would make the spec describe less than the code it pins — and the same identifiers now appear in `isLambdaUpdateInProgressError`, which asserts on `error.name === "ResourceConflictException"` precisely so the behaviour is tied to the AWS contract rather than to English message text.

Resolving rather than silently ignoring, so the thread is owned rather than left open.

File: platform/app/src/optimization_studio/server/lambda/__tests__/getProjectLambdaArn.test.ts [hunk 3]
```diff
@@ -93,7 +116,7 @@ describe("getProjectLambdaArn", () => {
   describe("ARN cache + single-flight", () => {
     /** @scenario First call hits AWS; subsequent calls within TTL serve from cache with zero AWS calls */
     it("serves repeated calls within TTL from cache with zero AWS calls", async () => {
-      const send = vi
+      const _send = vi
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply platform/app/src/optimization_studio/server/lambda/__tests__/getProjectLambdaArn.test.ts:120]: Fixed in b35adf3. The nested `it()` was extracted to a sibling, the `_send`/`send` bindings were unified, and the error-discrimination cases now seed a local `Timeout: 300` against a desired 900 so they actually reach the conflict handler instead of dying in readiness polling — which is exactly the third bullet here. The suite now collects and runs 19 tests with 0 collection errors.

File: platform/app/src/optimization_studio/server/lambda/__tests__/getProjectLambdaArn.test.ts [hunk 4]
```diff
@@ -270,4 +297,265 @@ describe("getProjectLambdaArn", () => {
...
+      expect(updates[0][0].input.FunctionName).toBe(
+        "langwatch_nlp-reconcile-env",
+      );
+      expect(updates[0][0].input.Environment.Variables).toEqual({
+        ...desiredEnvVars,
+        SOME_OTHER_VAR: "keep-me",
+      });
+    });
+
+    describe("error discrimination", () => {
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply platform/app/src/optimization_studio/server/lambda/__tests__/getProjectLambdaArn.test.ts]: Moot as of b35adf3 — the `describe("error discrimination")` block no longer exists. Four identically-bodied copies of that block had been duplicated in the file; they were collapsed to one and the surviving block sits under an existing BDD condition. Verified the collapse was a dedup and not a deletion: `grep -oE 'it\("[^"]+"' | sort | uniq -c` shows the three affected names going 4 -> 1 with the unique-name set unchanged at 18, and the 14 lines following each of the 4 occurrences hash identically per name.

File: platform/app/src/optimization_studio/server/lambda/__tests__/getProjectLambdaArn.test.ts [hunk 5]
```diff
@@ -270,4 +297,265 @@ describe("getProjectLambdaArn", () => {
...
+              ImageUri:
+                "123456789012.dkr.ecr.us-east-1.amazonaws.com/test:latest",
+            },
+          })
+          .mockRejectedValueOnce({
+            name: "ResourceConflictException",
+            message: "An update is in progress",
+          });
+
+        const arn = await getProjectLambdaArn("error-by-name");
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply platform/app/src/optimization_studio/server/lambda/__tests__/getProjectLambdaArn.test.ts:488]: Declining as out of scope for this PR, with reasoning rather than silence. The positional signature is pre-existing, not introduced here: it is present at the merge base (`git show 2174c30a46:.../lambda/index.ts` lines 401-403) and `git diff 2174c30a46..HEAD -- index.ts` shows the signature untouched by this branch. Converting it to `getProjectLambdaArn({ projectId })` is a product API change with two non-test callers (`server/nlpgo/nlpgoFetch.ts:153` and `lambda/index.ts:790`) on a PR whose subject is an NLP code-block timeout fix. You tagged it Heavy lift yourself. Worth doing as its own change; not folded in here.

File: platform/app/src/optimization_studio/server/lambda/__tests__/getProjectLambdaArn.test.ts [hunk 6]
```diff
@@ -270,4 +297,265 @@ describe("getProjectLambdaArn", () => {
...
+          })
+          .mockRejectedValueOnce(new Error("Access Denied"));
+
+        await expect(getProjectLambdaArn("error-unrelated")).rejects.toThrow(
+          "Access Denied",
+        );
+      });
+    });
+
+    /** @scenario A Lambda still on the old 1024 MB default is raised to 2048 */
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply platform/app/src/optimization_studio/server/lambda/__tests__/getProjectLambdaArn.test.ts:536]: Declining, with reasoning. `@scenario` is an established convention in this repo rather than decoration — 6908 occurrences across `platform/app/src` (e.g. `server/prompt-config/__tests__/prompt-tags.integration.test.ts:144`). This specific comment also carries a fact the test name does not: the test is named "raises a drifted MemorySize to the desired value", while the comment records that the drifted value is the old 1024 MB default being raised to 2048. That is added context, not a restatement.

File: platform/app/src/optimization_studio/server/lambda/__tests__/getProjectLambdaArn.test.ts [hunk 7]
```diff
@@ -270,4 +297,437 @@ describe("getProjectLambdaArn", () => {
...
+          Code: { ImageUri: "123456789012.dkr.ecr.us-east-1.amazonaws.com/test:latest" },
+        })
+        .mockRejectedValueOnce(
+          new Error("Access Denied"),
+        );
+
+      await expect(getProjectLambdaArn("error-unrelated")).rejects.toThrow(
+        "Access Denied",
+      );
+    });
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply platform/app/src/optimization_studio/server/lambda/__tests__/getProjectLambdaArn.test.ts]: **Moot at the current head (`9150117b`).** This finding lists duplicate suites to delete at lines 548-601, 622-675 and 678-731, but the file is now **571 lines total** — those ranges are past EOF. The duplication was already removed in `b35adf3`.

Evidence:

```
$ wc -l < platform/app/src/optimization_studio/server/lambda/__tests__/getProjectLambdaArn.test.ts
571
```

The three suites were deduplicated, not deleted — verified by multiplicity, not a set-difference. Each of the three test names appeared 4 times before and 1 time after, and the 14-line body following every occurrence hashed identically across all four copies (`44d0d97b88fd` x4, `54dd5a720f81` x4, `6270cbf2379f` x4). Unique test names went 18 -> 18; total `it(` went 28 -> 19, which is exactly 28 - (3 names x 3 redundant copies) = 19.

The substantive half of this finding — that the conflict tests were consuming the rejection in `pollLambdaUntilReady` rather than on the update command — was also fixed in `b35adf3`: the tests now force configuration drift with a local `Timeout: 300`, reject on `UpdateFunctionConfigurationCommand`, and supply a `.mockResolvedValue()` fallback for the readiness poll.

```
$ pnpm vitest run src/optimization_studio/server/lambda/__tests__/getProjectLambdaArn.test.ts
Test Files  1 passed (1)
     Tests  19 passed (19)
```

CI agrees: all four `test-unit` shards are green at this head, and `verify-ci-shard-tally.sh` confirms each printed a real tally line (no hard-floor forgery).

File: platform/app/src/optimization_studio/server/lambda/__tests__/clampCodeBlockTimeoutSeconds.test.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
+import { describe, it, expect, beforeEach } from "vitest";
+import { clampCodeBlockTimeoutSeconds } from "../index";
```
[the remark this answers is not in the document]
Comment (drewdrewthis) [reply platform/app/src/optimization_studio/server/lambda/__tests__/clampCodeBlockTimeoutSeconds.test.ts]: **Done in `1725a8b`.** Renamed via `git mv` (100% similarity, no content change):

```
$ git diff --stat -M 9150117b..1725a8bc
 ...eoutSeconds.test.ts => clampCodeBlockTimeoutSeconds.unit.test.ts | 0
```

The convention check supports it — the bare suffix is the minority in this codebase:

```
$ find src -name "*.unit.test.ts" | wc -l          -> 1776
$ find src -name "*.integration.test.ts" | wc -l   ->  483
$ find src -name "*.test.ts" ! -name "*.unit.test.ts" ! -name "*.integration.test.ts" | wc -l -> 210
```

Still collected and still green after the rename:

```
$ pnpm vitest run src/optimization_studio/server/lambda/__tests__/
Test Files  3 passed (3)
     Tests  42 passed (42)
```

The renamed file alone runs its 12 cases.

File: platform/app/src/optimization_studio/server/lambda/index.ts [hunk 9]
```diff
@@ -475,40 +752,11 @@ const resolveProjectLambdaArn = async (
       }
     }
   } else {
-    // Get complete function details to check image URI
-    const getFunctionCommand = new GetFunctionCommand({
-      FunctionName: functionName,
+    lambdaConfig = await syncExistingProjectLambda(lambda, functionName, {
```
Comment (langwatch-agent) [inline platform/app/src/optimization_studio/server/lambda/index.ts:801]: **P2 · Cached ARNs bypass the new reconciliation**

- **What you are looking at.** This synchronization only runs after a cache miss. The cluster-wide Redis ARN cache returns before `resolveProjectLambdaArn` whenever the image URI matches, and it stores only `{ arn, imageUri }`, not the desired environment, memory, or timeout.

- **Why it is scary.** A rollout that changes only `NLPGO_ENGINE_CODE_BLOCK_TIMEOUT_SECONDS` does not repair pre-existing functions on their next Studio request as the PR promises. The old shared cache survives the worker rollout and bypasses this call for up to ten minutes, leaving those functions on their old 60-second engine ceiling during the window. Include a desired-configuration fingerprint in the cache entry, or invalidate the entry when the desired Lambda configuration changes, so timeout-only deploys actually reach reconciliation.

**Review by langwatch-agent — COMMENTED** [review]
Static review found one P2. Details are inline. I inspected the diff and existing CI results; the local worktree has no installed dependencies, so I did not run the focused TypeScript or Helm suites.

LangWatch-Review: verdict=findings sha=be0c5c3a8e5cf16560912937900efc90c04284d9 p0=0 p1=0 p2=1 p3=0

**Review by rogeriochaves — DISMISSED** [review]

**Comment by langwatch-agent** [comment]
## Human Review Brief

> [!CAUTION]
> The author already names it: this raises a client-side network deadline from **120s to 630s**, so a genuinely hung code block occupies a scenario worker for about ten minutes instead of two. It is bounded by the engine's 600s ceiling and Lambda's 900s `Timeout`, so it cannot run away. The question to push on is capacity: how many workers exist, and what happens to everyone else's runs while several are held for ten minutes each.

| | |
|---|---|
| **Issue** | None linked. Combines `#7643`, `#7644`, `#7645`, `#7646` into one review surface |
| **Size** | +1634/-154 across 25 files |
| **Mode** | Targeted, on one new file |
| **Blast radius** | Every scenario run that executes a code block |
| **Author** | drewdrewthis |

<details>
<summary><b>What you are looking at, if the scenario execution path is not your directory</b></summary>

A scenario that runs code hands it to **nlpgo**, the Go workflow engine, and waits. Between the two sits a chain of deadlines, and the bug is that they were never reconciled: the client gave up at ~120s while the engine was still happily working, so a run died with the engine's answer still in flight.

The deadlines in play, outermost first: Lambda's `Timeout` (900s) bounds the whole invocation, the engine's own ceiling (600s) bounds a block, and the client's network deadline bounds how long we wait for the engine. Any client deadline shorter than the engine ceiling means the client can abandon work the engine will finish and bill for.

This also sits directly on top of `#7614`, which wired up the engine-side `NLPGO_ENGINE_*` timeout knobs that had been parsed and read by nothing. Those two together are what make the ceiling real on both ends.
</details>

<details>
<summary><b>Where the review goes</b></summary>

`platform/app/src/server/nlpgo/timeouts.ts` is new, 233 lines, and the author says every number in the PR derives from it. That is the right structure and it means the review is mostly one file:

1. **The ordering invariant.** Client deadline ≥ engine ceiling ≥ per-node timeout, and the whole chain under Lambda's 900s. If that ordering is asserted in code rather than maintained by hand, this class of bug cannot return. If it is only documented, it will drift again.
2. **The 630s specifically.** It is 600s plus 30s of slack. Confirm 30s covers what it needs to (dispatch, network, response serialisation) and that nothing else adds latency inside the same budget.
3. **What a hung block costs.** How many scenario workers exist, and is there any per-run or per-project cap on how many can be held simultaneously? Ten minutes is fine for one; it is an outage if a customer can fill the pool with them.
4. **Whether the client cancels.** When the client does eventually time out, does it tell the engine to stop, or does the engine keep working on an answer nobody will read? The second wastes the ceiling twice.
</details>

<details>
<summary><b>Standards note</b></summary>

Combining four PRs into one review surface, with a "scariest thing" line and a named file to start with, is the right call for a change whose whole point is that the numbers must agree with each other. Reviewing them separately would have meant nobody saw the chain.
</details>

> [!NOTE]
> Advisory only. The hound does not review code line by line, does not approve and does not merge.

File: platform/app/src/optimization_studio/server/lambda/index.ts (hunk 9 above)
Comment (drewdrewthis) [reply platform/app/src/optimization_studio/server/lambda/index.ts:801]: Good catch — this was a real gap and it is now fixed in d7e60c376e.

The cache compared only `image_uri`, which was correct until this PR gave a Lambda a *second* reason to need updating. A config-only rollout (raising `NLPGO_ENGINE_CODE_BLOCK_TIMEOUT_SECONDS` with no new image) left warm entries valid, so `getProjectLambdaArn` early-returned and skipped `reconcileProjectLambdaConfig` for up to the full 10-minute TTL.

Cache entries now also carry a `configFingerprint` built from `buildDesiredLambdaEnvironmentVariables()` — the same call the reconcile drift check uses, so the cache key and the drift check cannot disagree — plus `MemorySize` and `Timeout`. Env keys are sorted before joining so key reordering can never cause a spurious miss.

Entries written before this change lack the field, compare unequal, and fall through to a reconcile: a one-time invalidation on rollout that fails open rather than serving stale. No new env vars.

Proven non-vacuous: reverting just the fingerprint comparison fails the new test `invalidates the cache when the desired timeout changes (no image change)` with `AssertionError: expected 3 to be greater than 3`. Full file 20/20 passing with the fix in place.

**Review by rogeriochaves — APPROVED** [review]
